# Lab 23: 2D Lists

First, **read Section 7.9** or this lab won't make much sense.

A 2D list in Python is really a list of lists.

Imagine we have list1: [1,2,3,4,5] and list2: [2,3,4,5,6] and list3: [3,4,5,6,7]

Now imagine we have a list, where each element in the list is itself a list. The list output would be: 

    [list1, list2, list3]
    
Now by substitution we get:

    [[1,2,3,4,5],[2,3,4,5,6],[3,4,5,6,7]]
    
And if we output the 2D list with some formatting to make the rows and columns cleaner, it'd look like:

    1   2   3   4   5
    2   3   4   5   6
    3   4   5   6   7

## List Comprehension to Create 2D Lists

To create a single 1D list of size 7 of all 0's, we'd use:

    [0 for i in range(7)]
    
Or rather than 0, we do a random.

Run this code for a demonstration.

In [1]:
import random
COLS = 7

# generate list of 0's
values = [0 for i in range(COLS)]
print(values)

# generate list of randoms
values = [random.randint(1, 10) for i in range(COLS)]
print(values)

[0, 0, 0, 0, 0, 0, 0]
[9, 7, 8, 3, 9, 10, 5]


We abstract this by one layer to **initialize a 2D list with all 0's** using list comprehension.

To create a separate list of 0's for the number of rows, we'd use:

    values = [0 for j in range(ROWS)]
    
Suppose that y equals this code: y = [0 for i in range(COLS)]

Now substitute y in the following code to get the full Python statement:

    values = [y for j in range(ROWS)]
    
    and if y = [0 for i in range(COLS)]
    
    then the statement after substitution is: 
    values = [[0 for i in range(COLS)] for j in range(ROWS)]
    
If we want to initialize with other values, we just put the value we want in place of the '0' in this code:

    values = [[0 for i in range(COLS)] for j in range(ROWS)]

For example, a random number 1-10:

    values = [[random.randint(1,10) for i in range(COLS)] for j in range(ROWS)]

## Exercise 23-1: Use List Comprehension to Initialize a 2D List to 0's

Create a 5x5 2D list initialized to 0's. Use list comprehension to do so. Output the list without formatting by using this simple for loop:

    for row in arr:
        print(Row)

Then do the same for another 5x5 list of random 0's or 1's.

Expected output:

    [0, 0, 0, 0, 0]
    [0, 0, 0, 0, 0]
    [0, 0, 0, 0, 0]
    [0, 0, 0, 0, 0]
    [0, 0, 0, 0, 0]

    [0, 0, 1, 0, 1]
    [1, 1, 1, 1, 0]
    [1, 1, 0, 0, 0]
    [1, 0, 0, 0, 0]
    [1, 0, 0, 0, 1]

[0, 0, 0, 0, 0]
[0, 0, 0, 0, 0]
[0, 0, 0, 0, 0]
[0, 0, 0, 0, 0]
[0, 0, 0, 0, 0]

[0, 0, 0, 0, 0]
[1, 0, 1, 0, 0]
[1, 0, 1, 0, 1]
[0, 1, 0, 0, 0]
[0, 1, 0, 1, 0]


## Traversing a 2D List

Now that we can create and initialize a 2D list to 0's, let's figure out how to traverse the list.

We do so using nested for loops. The outermost for loop traverses the rows, while the innermost for loop traverses the columns. 

It's analagous to a clock: the outer for loop is the hours, and the inner for loop is the minutes. The  minutes for loop has to spin 60 times. Then the hour loop is incremented and the minutes loop restarts. Here's an example:

    hour  minute
    02    57   <-- inner minutes loop spins faster
    02    58
    02    59   <-- at the end of the loop, increment hours and reset
    03    00
    03    01
    
Typically we store the number of rows and columns in a 2D list as CONSTANTS, for easy reference.

The format of the nested for loops is:

    for r in range(ROWS):
        for c in range(COLS):
            # inside the loop, access the element with arr[r][c]

## Exercise 23-2: Combination Hacker

A certain combination lock has 2 dials, and each dial can only be the numbers 1-4. Write a program to enumerate all possible combinations. Use two nested for loops (one for each dial), each running 4 times. Expected output:

    11
    12
    13
    14
    21
    22
    23
    24
    31
    32
    33
    34
    41
    42
    43
    44

11
12
13
14
21
22
23
24
31
32
33
34
41
42
43
44


## Exercise 23-3: Write a Function to Print a Formatted 2D List

You still have the 2D list 'arr' in memory from the code above.

Write a function to print a formatted 2D list. Name your function 'print2D(). It will receive a 2D list as a parameter. It will have to step through the 2D list element by element and print them in a well-formatted table. **Make your field width 6 for good readability.** Use your print(f'{}') statements for this.

Note you can't use the same simple for loop that we used in the code above, because that output is unformatted.

Test your function by sending your 2D list of 0's to it and get the formatted output shown here:

     0     0     0     0     0
     0     0     0     0     0
     0     0     0     0     0
     0     0     0     0     0
     0     0     0     0     0

     0     0     0     0     0
     0     0     0     0     0
     0     0     0     0     0
     0     0     0     0     0
     0     0     0     0     0


## Exercise 23-4: Write Randoms to File

Write a program that will write 120 random integers in the range (10, 200) to an external file named 'data.txt'. Program 6-6 in the textbook might be a refresher.

## Exercise 23-5: Read External Values Into 2D List

Write a program that reads the 120 integers in 'data.txt' and stores them in a 10x12 2D list. Program 6-7 in the textbook might be a refresher. Use your print2D() function to output the list. 

   102    23   125   162    41   135    94   122    76   153    96   106
    76    48   179   108   180    62    48   136   197   153    62   150
    24    90    52    93   142   195   199   154   197    43    65    22
    69   103   186    23   190   183    92   191   178    37   135    86
   101   184    32    38   150    68   197    85    88    89    74    15
    47   124   139    58    36   105    64    32    60   119    81    48
    99   107    30    38    48    65    81    72    58    35   139   156
   172   137    97   200    29    66   114   116    49   161   174   120
   125   170   136    17   199    53    36   183   143    31   164   150
    62   123    45   157   116   117   100   154   148   194   150    83


## Practical Application: Dog Trainer

Suppose you're a dog trainer. You want to keep track of each dog's skill test scores. 

If you had one dog, the list might look like this: [88,92,68,50,100]

Another dog's list might be: [95,94,77,90,87] and another's [100,100,50,88,98].

If we stack these lists, the rows would represent which dog it is (perhaps it's Dog #0, Dog #1, or Dog #2 based on what row their data is on). The columns represent each individual dog's skill test scores. 

So the rows dimension is dogs; the column dimension is scores. Creating the 2D list looks like:

     88   92   68   50  100
     95   94   77   90   87
    100  100   50   88   98

We could sum the rows, sum the columns, total the whole list, etc.

### Summing Rows

Here's the sleekest way I know of summing the rows. We could code this manually, using an accumulator in that inner for loop. But this way is better and it does the trick in one line of code.

    row_totals = [sum(r) for r in list]

This returns a list and assigns it to 'row_totals'. 

The list comprehension says, "for every row in list, use the sum() function on that row." We can use sum() here because since sum() takes a list as an argument, we can give the variable 'r' as that argument since 'r' represents one of the inner lists.

In [7]:
DOGS, TESTS = 3, 5
def main():
    dog1 = [88,92,68,50,100]
    dog2 = [95,94,77,90,87]
    dog3 = [100,100,50,88,98]
    scores = [dog1, dog2, dog3]
    print2D(scores)
    
    # calculate row sums
    row_totals = [sum(r) for r in scores]
    for i in range(DOGS):
        print(f'Row {i} sum: {row_totals[i]}')

def print2D(lst):
    for row in lst:
        for col in row:
            print(f'{col:6d}', end='')
        print()
    print()
    
main()

    88    92    68    50   100
    95    94    77    90    87
   100   100    50    88    98

Row 0 sum: 398
Row 1 sum: 443
Row 2 sum: 436


### Summing Columns

Summing columns is a bit tougher, so we use a little trick: we just reverse the traditional order of the indices 'row' and 'col' in the for loop structure. COL will come first, then ROW. We use an accumulator variable inside. It works, try it!

In [8]:
DOGS, TESTS = 3, 5
def main():
    dog1 = [88,92,68,50,100]
    dog2 = [95,94,77,90,87]
    dog3 = [100,100,50,88,98]
    scores = [dog1, dog2, dog3]
    print2D(scores)
    
    # calculate row sums
    row_totals = [sum(r) for r in scores]
    
    # print row sums
    for i in range(DOGS):
        print(f'Row {i} sum: {row_totals[i]}')
    print()

    # calculate and print column sums
    for t in range(TESTS):
        col_sum = 0  
        for d in range(DOGS):
            col_sum += scores[d][t]
        print(f'Column {t} sum: {col_sum}')

def print2D(lst):
    for row in lst:
        for col in row:
            print(f'{col:6d}', end='')
        print()
    print()
    
main()

    88    92    68    50   100
    95    94    77    90    87
   100   100    50    88    98

Row 0 sum: 398
Row 1 sum: 443
Row 2 sum: 436

Column 0 sum: 283
Column 1 sum: 286
Column 2 sum: 195
Column 3 sum: 228
Column 4 sum: 285


## Exercise 23-6: Sum Rows & Columns

First, using **list comprehension**, generate a 8x10 list of random values (range: 0, 500). We're not reading from the file for this exercise; we're using list comprehension that we looked at earlier.

Output the list using your print2D() function.

Then output the row and column sums. To do so, you will have had to have stored the sums in separate lists that you can now access. You already have code that will sum the rows; the syntax is (this is from the cell above):

    row_totals = [sum(r) for r in scores]

To **create the list of column sums**, first create a list of size COLS named 'sum_cols'. Then check out this code block:

    # calculate column sums
    for c in range(COLS):
        for r in range(ROWS):
            sum_cols[c] += arr[r][c]


Output these two lists in this format. Note that there are only 10 rows, so the last two fields on that 'Row sum' row are empty:

                0    1    2    3    4    5    6    7    8    9   10   11
    Row sum: 1397 1449 1477 1304 1034 1100 1129 1049 1122  954
    Col sum: 1094  668 1091  970  950  895 1107 1052 1012  984 1326  866

Then calculate and output statistics:

    ***STATISTICS***
        Grand total: 12015
        Average value: 100.125
        Max value: 186
        Min value: 18

**max(lst)** is a function that will **return the maximum element** in the list 'lst'. Experiment with **nesting this function** in order to return the maximum value from the 2D list. Also code minimum. 

Sample output:

       182   278    66   312   175   268    30   460     9   380
       148   490   171   390   237   269   370   184   221   320
       361    92   175   288   117   240   343   238    39   459
       252   361   385   388   351   299   297   450   190    73
       348   302   247   489   427   146   293    39   288   336
       482   209    57   267   409    57   177   163   462   358
       309   373   326   388   408   181   196   355   199   432
        12   227   268   259    80    73   262   374   339   460

                0    1    2    3    4    5    6    7    8    9
    Row sum: 2160 2800 2352 3046 2915 2641 3167 2354
    Col sum: 2094 2332 1695 2781 2204 1533 1968 2263 1747 2818

    ***STATISTICS***
    Grand total: 21435
    Average value: 267.9375
    Max value: 482
    Min value: 12

   133   367   410   476    74    41    83   492   104    99
   196   423   487   185   480   370    94   415    89   431
   419   282    64    17    80   356   344   319   429   179
   434   349   429    98   110   193   178    61    90   195
    66   388   113   108   227   236   190   142   321   434
    86   290   438   434   120   145   178   171   203   262
   158   197     4   312    11   335   310    53   431   207
   455   465   315   386    15   443   241   456   289   164

            0    1    2    3    4    5    6    7    8    9
Row sum: 2279 3170 2489 2137 2225 2327 2018 3229
Col sum: 1947 2761 2260 2016 1117 2119 1618 2109 1956 1971

***STATISTICS***
Grand total: 19874
Average value: 248.425
Max value: 465
Min value: 66
